src/System.Runtime.WindowsRuntime/tests/System/IO/AsWinRTStreamTests.cs

```diff
@@ -0,0 +1,413 @@
+
+            IAsyncOperationWithProgress<IBuffer, uint> readOp = stream.ReadAsync(s_buffer, (uint)TestStreamProvider.ModelStreamLength, inputStreamOptions);
+
+            if (completesSynchronously)
+            {
+                // New readOp for a stream where we know that reading is sycnhronous must have Status = Completed
+                Assert.Equal(AsyncStatus.Completed, readOp.Status);
+            }
+            else
+            {
+                // Note the race. By the tie we get here, the status of the op may be started or already completed.
+                AsyncStatus readOpStatus = readOp.Status;
+                Assert.True(readOpStatus == AsyncStatus.Completed || readOpStatus == AsyncStatus.Started, "New readOp must have Status = Started or Completed (race)");
+            }
+
+            s_progressCallbackInvoked = false;
+            s_completedCallbackInvoked = false;
+
+            uint readOpId = readOp.Id;
+            EventWaitHandle waitHandle = new ManualResetEvent(false);
+
+            readOp.Progress = (asyncReadOp, bytesCompleted) =>
+            {
+
+                s_progressCallbackInvoked = true;
+
+                // asyncReadOp.Id in a progress callback must match the ID of the asyncReadOp to which the callback was assigned
+                Assert.Equal(readOpId, asyncReadOp.Id);
+
+                // asyncReadOp.Status must be 'Started' for a asyncReadOp in progress
+                Assert.Equal(AsyncStatus.Started, asyncReadOp.Status);
+
+                Assert.True(0 <= bytesCompleted && bytesCompleted <= (uint)TestStreamProvider.ModelStreamLength,
+                    "bytesCompleted must be in range [0, maxBytesToRead] asyncReadOp in progress");
+            };
+
+            readOp.Completed = (asyncReadOp, passedStatus) =>
+            {
+                try
+                {
+                    s_completedCallbackInvoked = true;
+
+                    // asyncReadOp.Id in a completion callback must match the ID of the asyncReadOp to which the callback was assigned
+                    Assert.Equal(readOpId, asyncReadOp.Id);
+
+                    // asyncReadOp.Status must match passedStatus for a completed asyncReadOp
+                    Assert.Equal(passedStatus, asyncReadOp.Status);
+
+                    // asyncReadOp.Status must be 'Completed' for a completed asyncReadOp
+                    Assert.Equal(AsyncStatus.Completed, asyncReadOp.Status);
+
+                    IBuffer resultBuffer = asyncReadOp.GetResults();
+
+                    // asyncReadOp.GetResults() must not return null for a completed asyncReadOp
+                    Assert.NotNull(resultBuffer);
+
+                    Assert.True(0 < resultBuffer.Capacity,
+                        "resultBuffer.Capacity when a completed asyncReadOp was expected to read more than zero bytes");
+
+                    Assert.True(0 < resultBuffer.Length,
```

Comment:
These can be done with Assert.InRange, which will report what the value was if it fails.